src/main/java/io/cdap/plugin/gcp/spanner/sink/SpannerOutputFormat.java

```diff
@@ -231,4 +243,125 @@ public void abortTask(TaskAttemptContext taskAttemptContext) {
+      }
+    } else {
+      Schema.Type type = unionSchema.getType();
+      switch (type) {
+        case BOOLEAN:
+          Collection<Boolean> listBoolean = toCollection(name, type.toString(), values);
+          builder.set(name).toBoolArray(listBoolean);
+          break;
+        case STRING:
+          Collection<String> listString = toCollection(name, type.toString(), values);
+          builder.set(name).toStringArray(listString);
+          break;
+        case LONG:
+          Collection<Long> listValue = toCollection(name, type.toString(), values);
+          builder.set(name).toInt64Array(listValue);
+          break;
+        case DOUBLE:
+          Collection<Double> listDouble = toCollection(name, type.toString(), values);
+          builder.set(name).toFloat64Array(listDouble);
+          break;
+        case INT:
+          Collection<Integer> integerList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toInt64Array(
+              integerList.stream()
+                .map(Integer::longValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case FLOAT:
+          Collection<Float> floatList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toFloat64Array(
+              floatList.stream()
+                .map(Float::doubleValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case BYTES:
+          Collection<byte[]> bytesList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toBytesArray(
+              bytesList.stream()
+                .map(ByteArray::copyFrom)
+                .collect(Collectors.toList())
+            );
+          break;
+        default:
+          throw new IllegalStateException("Array of '" + type.name() + "' type currently not supported.");
+      }
+    }
+  }
+
+  private static <T> Collection<T> toCollection(String fieldName, String fieldType, Object value) {
+    Function<String, Collection> valueExtractor = name -> {
+      throw new UnexpectedFormatException(
+        String.format("Field '%s' of type '%s' has unexpected value '%s'", name, fieldType, value));
+    };
+
+    if (value instanceof Collection) {
+      valueExtractor = name -> (Collection<?>) value;
+    } else if (value.getClass().isArray()) {
+      valueExtractor = name -> convertToObjectCollection(value);
+    }
+    return getValue(valueExtractor, fieldName, fieldType, Collection.class);
+  }
+
+  private static Collection<Object> convertToObjectCollection(Object array) {
+    Class ofArray = array.getClass().getComponentType();
+    if (ofArray.isPrimitive()) {
+      List<Object> list = new ArrayList<>();
+      int length = Array.getLength(array);
+      for (int i = 0; i < length; i++) {
+        list.add(Array.get(array, i));
+      }
+      return list;
+    } else {
+      return Arrays.asList((Object[]) array);
```

Comment:
how come we need to check if the component is primitive? Does it work if we always do this?

Reply:
If we try to convert array of primitives to Object[] we will get an error.